neural_lam/models/ar_model.py

```diff
@@ -108,25 +110,15 @@ def __init__(
         self.grid_dim = (
             2 * self.grid_output_dim
             + grid_static_dim
-            + num_forcing_vars
+            # Factor 2 because of temporal embedding or windowed features
+            + 2
+            * num_forcing_vars
             * (num_past_forcing_steps + num_future_forcing_steps + 1)
```

Comment:
Could you explain more why there is a factor 2 here? I understand that there is an additional temporal embedding variable (time delta between state and forcing), but would that not just be `num_forcing_vars + 1`? What am I missing? The different forcing variables should not have different time deltas so you need twice as many vars?

Reply:
Oh right, it's more thatn `+ 1` since there is the windowing as well. But then I would expect `(num_forcing_vars + 1) * (num_past_forcing_steps + num_future_forcing_steps + 1)` rather than the `* 2`.

Reply:
You are right, all forcing/boundary features share the same temporal embedding. I fixed this in https://github.com/mllam/neural-lam/pull/90/commits/16d5d04bbd9e49a1fe53e56ff95e22d565326c67 and updated the time slicing test to cover also forecast-type data: https://github.com/mllam/neural-lam/pull/90/commits/8fd7a10fc1d01c998641baf228dfa66f8630280c @leifdenby you were also asking for this

PS: What is in terms of memory/computational efficiency the best way to return such a tensor that repeats over 2 dimensions `time, grid_index`? Should we enforce `signed Integers` here for improved memory management, even though the `time_step_diff` can be `float`?

Reply:
Nice! Should the lines I commented on in `ar_model.py` also be updated then?

Regarding memory management: You can have a tensor with large dimensions that share the same memory, e.g. for all grid points. If you construct the tensor on GPU you can achieve this with https://pytorch.org/docs/stable/generated/torch.Tensor.expand.html (and I think also other ways). But I am unsure what happens when you create such a tensor on CPU and then pass it o GPU. Maybe something that should be tested (e.g. by checking that memory is still shared when it's on the GPU by doing some in-place operation).

Reply:
Great!

Reply:
@sadamov Does this actually work correctly now? I am getting dimension mismatches when building MLPs with the dimensions as given after https://github.com/mllam/neural-lam/commit/252a33cd5903aca793f480e871912cc8b8616df2. In `_process_windowed_data` I am going from 6 forcing features (2 features windowed over 3 times) to 12 features. If we only include the delta time once per time in the window this should be 6 features to 6+3 = 9.

Reply:
This now works and is extensively tested using new and old tests:
Tests: https://github.com/mllam/neural-lam/pull/90/commits/1d53ce7b86ee8c936a1c8c2fd9bad58bb672b844
Fixing slicing/embedding: https://github.com/mllam/neural-lam/pull/90/commits/cfe1e278ae16a4ec8e19dc3f2db79e976e8014d6

Reply:
Now it seems correct, yes :smile: